src/adapter/src/catalog.rs

```diff
@@ -3698,39 +3723,113 @@ impl<S: Append> Catalog<S> {
+        let mut names = vec![];
+        let mut instance_ops = vec![];
+        for instance_name in instance_names {
+            let instance = self
+                .resolve_compute_instance(instance_name)
+                .expect("instance name already validated");
+            for replica_name in instance.replica_id_by_name.keys() {
+                names.push((instance_name.clone(), replica_name.clone()));
+            }
+            instance_ops.push(Op::DropComputeInstance {
+                name: instance_name.into(),
+            });
+        }
+        let (ids, mut ops) = self.drop_compute_instance_replica_ops(&names);
+        ops.extend(instance_ops);
+        (ids.into_iter().into_group_map(), ops)
+    }
+
+    /// Creates the catalog operations to drop the given compute instance
+    /// replicas.
+    ///
+    /// Returns the (instance ID, replica ID) pairs to drop, and the catalog
+    /// operations to do so.
+    pub fn drop_compute_instance_replica_ops(
+        &self,
+        names: &[(String, String)],
+    ) -> (Vec<(ComputeInstanceId, u64)>, Vec<Op>) {
+        let mut ops = vec![];
+        let mut replicas_to_drop = vec![];
+        let mut ids_to_drop = vec![];
+        for (instance_name, replica_name) in names {
+            let instance = self
+                .resolve_compute_instance(instance_name)
+                .expect("instance name already validated");
+            ops.push(Op::DropComputeReplica {
+                name: replica_name.clone(),
+                compute_name: instance_name.clone(),
+            });
+            let replica_id = instance.replica_id_by_name[replica_name];
+
+            // Determine from the replica which additional items to drop. This is the set
+            // of items that depend on the introspection sources. The sources
+            // itself are removed with Op::DropComputeReplica.
+            let logging = &instance
+                .replicas_by_id
+                .get(&replica_id)
+                .unwrap()
+                .config
+                .logging;
+
+            let log_and_view_ids = logging.source_and_view_ids();
+            let view_ids: HashSet<_> = logging.view_ids().collect();
+
+            for log_id in log_and_view_ids {
+                // We consider the dependencies of both views and logs, but remove the
+                // views itself. The views are included as they depend on the source,
+                // but we dont need an explicit Op::DropItem as they are dropped together
+                // with the replica.
+                ids_to_drop.extend(
+                    self.get_entry(&log_id)
+                        .used_by()
+                        .into_iter()
+                        .filter(|x| !view_ids.contains(x)),
+                )
+            }
+
+            replicas_to_drop.push((instance.id, replica_id));
+        }
+
+        ops.extend(self.drop_items_ops(&ids_to_drop));
+        (replicas_to_drop, ops)
+    }
```

Comment:
These are essentially just code movement from `sequence_drop_compute_instance` and `sequence_drop_compute_replica` in the coordinator.